File: javascript/18_testing.js
// 18 testing - tiny self-check (no jest required)

function add(a, b) {
  return a + b;
}

function divide(a, b) {
  if (b === 0) throw new Error("cannot divide by zero");
  return a / b;
}

function assert(cond, msg) {
  if (!cond) throw new Error(msg || "assert failed");
}

assert(add(2, 3) === 5, "add 2+3");
assert(add(-1, 1) === 0, "add -1+1");
assert(divide(10, 2) === 5, "divide");

try {
  divide(1, 0);
  throw new Error("should have thrown");
} catch (e) {
  if (e.message !== "cannot divide by zero") throw e;
}

console.log("all tests passed");

File: python/18_testing.py
# 18 testing
# run: python -m pytest python/18_testing.py -q
# or just run this file for a tiny self-check

def add(a, b):
    return a + b

def divide(a, b):
    if b == 0:
        raise ValueError("cannot divide by zero")
    return a / b

# tiny self-test without pytest dependency
def test_add():
    assert add(2, 3) == 5
    assert add(-1, 1) == 0

def test_divide():
    assert divide(10, 2) == 5
    try:
        divide(1, 0)
        assert False, "should have raised"
    except ValueError:
        pass

if __name__ == "__main__":
    test_add()
    test_divide()
    print("all tests passed")
